### DISCUSSION TITLE
Format before fstab?

### DISCUSSION THREAD

**Discussion by hongkongkiwi** [finit-project/finit#258] [Q&A]
I wanted to get your advice.

Before mounting an emmc I wanted to format the particular partitions (if they are not formatted). The emmc is totally black. At the moment, the best way I can see to do this from finit is to call a script from fstab to check and format before the mount takes place using the same method as we use for mkdir#-p

Does that make sense to you? 

Am I right in assuming that the only way to run something before fstab is a plugin in finit? (is it even possible to run a plugin before mounting fstab?)

**Comment by hongkongkiwi** [comment]
**Side question:** What is the behaviour if the "run" command in /etc/fstab.conf has a condition? will no futher run commands fire as the first run is blocked.

**Reply by troglobit** [reply]
> What is the behaviour if the "run" command in /etc/fstab.conf has a condition? will no futher run commands fire as the first run is blocked.

Sorry, I don't understand the question. The mount helper commands, e.g. `mkdir -p`, in fstab.conf are executed by BusyBox mount, not Finit.

**Comment by troglobit** [comment]
> Before mounting an emmc I wanted to format the particular partitions (if they are not formatted). The emmc is totally black. At the moment, the best way I can see to do this from finit is to call a script from fstab to check and format before the mount takes place using the same method as we use for mkdir#-p
> Does that make sense to you?

Yes, this would be the recommended way to do this at the moment.

> Am I right in assuming that the only way to run something before fstab is a plugin in finit? (is it even possible to run a plugin before mounting fstab?)

Yes, that's what we did at my old job. We had C plugins for the various "hook points" that exist in Finit.  There are a few that may be of use to you: `HOOK_BANNER`, `HOOK_ROOTFS_UP`, and `HOOK_MOUNT_ERROR`.  The latter is what we used when mounting our read-write JFFS2 partition failed we logged the error, reformatted the partition and rebooted.

**Reply by troglobit** [reply]
**Update:** my partner in crime, @wkz, just pushed PR #262, which you may be interested in @hongkongkiwi :-) 

It's a way to call run-parts scripts on the various hook points in Finit.  Only intended for debug purposes, so it will not get any official documentation, but you should be able to make out how it works from the PR.

**Reply by hongkongkiwi** [reply]
Thank you very much for this, that will be very helpful.
